# Metagenomics workflow

This repository includes the bioinformatic workflow for metagenomics, all parameters are set to 
work properly on specific data from our lab. Before any usage, please check if it fits your data as well. 

## üöÄ Quick usage

Whole workflow can be executed by main_metagenomics.sh, where you have to change paths to conda environments and to your data

```bash
bash main_metagenomics.sh
```

### üì• Inputs

### üì§ Outputs

### üì¶ Requirements
databases.csv
adapters.fasta
prebuilt metaphlan dbs
prebuilt kraken dbs
phix bowtie indexes
fastqc
multiqc
hostile
fastp
metaphlan
kraken
bracken
nextflow

## ‚ÑπÔ∏è About
The workflow consists of multiple modules, all will be exececuted if not set otherwise:

---
### üßπ **<u>1. Quality control</u>**

Runs **FastQC & MultiQC** to provide quality control report of all samples. Additionally, it creates a **QUICK SUMMARY** 
of adapter content, sequencing depth, overrepresented sequences and their BLAST hits, for overall summary of the run.


**Input**: raw fastqcs  
**Output**:
- *quality_raw* folder with fastqc and multiqc data
- *run_info* folder with:
     - raw_custom_summary.txt - adapter content, sequencing depth, overrepresented sequences
     - raw_multiqc_report.html - copy of the multiqc_report.html
     - reads_count_summary.txt - read counts
---    


### ‚öôÔ∏è **<u>2. Preprocessing</u>**

Runs **fastp** for preprocessing the raw fastq files. It performs quality filtering and trimming. Generally, it is executed with the default parameters with the minor changes. See the command below.

```bash
fastp \
    -i sample_R1_001.fastq.gz \
    -I sample_R2_001.fastq.gz \
    -o /path/to/output/sample_R1_trimmed.fastq.gz \
    -O /path/to/output/sample_R2_trimmed.fastq.gz \
    --adapter_fasta adapters.fasta \
    --trim_poly_g --trim_poly_x \
    --cut_tail --cut_front \
    --length_required 75 \
    --thread 5 \
    --html ${path_output}/reports/{/}.html \
    --json ${path_output}/reports/{/}.json"
```

Next, it runs **hostile** to remove human and phix contamination. As for human decontamination, it uses hostile's prepared indexes (human-t2t-hla-argos985-mycob140) that consist of T2T-CHM13v2.0 + IPD-IMGT/HLA v3.51 masked with 150mers for 985 FDA-ARGOS bacterial & 140 mycobacterial genomes. As for phix decontamination, it uses custom indexes of [phiX174](https://www.ncbi.nlm.nih.gov/nuccore/9626372), built by:

```bash
bowtie2-build phiX174.fasta phiX174
```

See the executed commands below. 

```bash
# human decontamination
hostile clean \
--fastq1 sample_R1_trimmed.fastq.gz \
--fastq2 sample_R2_trimmed.fastq.gz \
--index human-t2t-hla-argos985-mycob140
--output /path/to/project_dir/decontaminated/human/

# phix decontamination
hostile clean \
--fastq1 sample_R1_trimmed.clean_1.fastq.gz \
--fastq2 sample_R2_trimmed.clean_2.fastq.gz \
--index path/to/prebuilt/phix_indexes \
--output /path/to/project_dir/decontaminated/human_phix/
```

**Input**: raw fastqcs  
**Output**: 
- *trimmed* folder - results of fastp
- *decontaminated* folder:
    - *decontaminated/human/* folder - results of hostile's human removal
    - *decontmainated/human_phix/* folder - results of hostile's phix removal
- *run_info* folder with:
     - read_counts_summary.txt - updated read counts track with <u>trimmed</u> and <u>decontaminated</u> columns. 
     - trimmed_multiqc_report.html - copy of the multiqc_report.html containing fastp and fastqc reports
---

### üß¨ **<u> 3. Taxonomic profiling</u>**

Performs taxonomic profiling using one of two (or both) tools: **Metaphlan 4** and/or **Kraken+Bracken**.

- **Metaphlan 4** uses a set of species-specific marker genes to identify and quantify microbial taxa. MetaPhlAn 4 dramatically expands its database by incorporating thousands of new species-level genome bins (SGBs), including both known (kSGBs) and unknown (uSGBs) taxa.

- **Kraken 2** assigns taxonomic labels by matching k-mers from reads to reference genomes using an exact-match k-mer hash strategy. **Bracken** builds on Kraken results to re-estimate species (or genus) abundances more accurately using a Bayesian method.

*Which to use*:

This pipeline is designed to offer two separate approaches that are independent of each other. For an objective selection of a suitable tool, see some of the benchmark studies, for example [LEMMI](https://lemmi.ezlab.org/). When not specified by user, the pipeline executes Metaphlan 4.  

We offer a brief overview and comparison of Kraken and Metaphlan (generated by Chatgpt 5) below.


***Summary Comparison***

| Feature                  | MetaPhlAn 4                                              | Kraken + Bracken                                         |
|--------------------------|----------------------------------------------------------|-----------------------------------------------------------|
| **Methodology**          | Marker gene mapping (phylogenetically curated)           | Whole-read k-mer matching with abundance re-estimation    |
| **Accuracy**             | High precision, balanced recall, very low false positives | High recall, but many low-abundance false positives       |
| **Abundance Estimation** | Accurate, especially for known and unknown SGBs          | Accurate when filtered, but may overestimate rare taxa     |
| **Speed / Resources**    | Moderate speed; depends on mapping coverage              | Extremely fast and resource-efficient                      |
| **Database Coverage**    | Extensive with many uncharacterized species              | Dependent on database; may miss underrepresented genomes   |
| **Best Use**             | When accuracy and specificity matter (e.g., marker-based profiling in complex samples) | Rapid, broad surveys where speed is crucial or followed by filtering |

---

### üî¨ **4. Functional profiling**     

üëÄ Comming soon üëÄ




